server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -346,6 +374,67 @@ public String toString() {
         }
     }
 
+    @Override
+    protected TransportWriteAction.AsyncAfterWriteAction createAsyncAfterWriteAction(
+        IndexShard indexShard,
+        WriteRequest<?> request,
+        Translog.Location location,
+        TransportWriteAction.RespondingWriteResult respond,
+        Logger logger
+    ) {
+        return new AsyncAfterShardBulkAction(
+            indexShard,
+            request,
+            location,
+            respond,
+            logger,
+            latestSuccessfulLeaderCheckerTimeSupplier,
+            latestLeaderCheckThreshold
+        );
+    }
+
+    private static class AsyncAfterShardBulkAction extends AsyncAfterWriteAction {
+
+        private final IndexShard indexShard;
+        private final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier;
+        private final TimeValue leaderCheckerTimeThreshold;
+
+        public AsyncAfterShardBulkAction(
+            final IndexShard indexShard,
+            final WriteRequest<?> request,
+            @Nullable final Translog.Location location,
+            final RespondingWriteResult respond,
+            final Logger logger,
+            final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier,
+            final TimeValue leaderCheckerTimeThreshold
+        ) {
+            super(indexShard, request, location, respond, logger);
+            this.indexShard = indexShard;
+            this.latestSuccessfulLeaderCheckerTimeSupplier = latestSuccessfulLeaderCheckerTimeSupplier;
+            this.leaderCheckerTimeThreshold = leaderCheckerTimeThreshold;
+        }
+
+        @Override
+        protected void finishOnNoPendingOps() {
+            if (indexShard.isRemoteTranslogEnabled() && elapsedTimeSinceLatestSuccessfulLeaderCheckBeforeThreshold()) {
```

Comment:
if the leader check time has not elapsed, but primary got assigned somewhere else?

Reply:
Lets consider following cases -
 - One ore more active replica - In such cases, primary term validation / no-op replication (status quo) would kick and fail the request and the underlying shard on realising it's primary term is lesser than the primary term seen by the replicas. 
 - Zero active replicas (0 configured replicas or last standing shard) - Since this shard does not have any other shards in it's replication group, it means if it gets isolated then the shard/index is red from cluster manager's point of view. Now, in such cases, this shard can continue to accept requests as long as the auto restore of red shard/index kicks in.